### PR TITLE
Second part of moving all cognito related methods

### PR DESCRIPTION
**Pull request by j-maynard**

### CODE REVIEW

File: lib/auth/authentication_backend.rb [hunk 1]
```diff
@@ -29,7 +24,130 @@ def authentication_flow(params)
...
+
+  # Returns a secret shared code to associate a TOTP app/device with
+  def associate_device(access_token:)
+    associate = client.associate_software_token(access_token: access_token)
+    associate.secret_code
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new("Error occurred associating device with error #{e.message}")
+  end
+
+  def enrole_totp_device(access_token:, totp_code:)
```
Comment (jakubmiarka) [inline lib/auth/authentication_backend.rb]:
```suggestion
  def enrol_totp_device(access_token:, totp_code:)
```

File: app/models/healthcheck/cognito_check.rb [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
     def name
       :cognito_connectivity
     end
 
     def status
-      SelfService.service(:cognito_client).describe_user_pool(
-        user_pool_id: Rails.configuration.cognito_user_pool_id
-      )
-      OK
+      status
```
Comment (jakubmiarka) [inline app/models/healthcheck/cognito_check.rb]: this is too generic, could we either namespace it call it something like `authentication_backend_status` ?

File: lib/auth/authentication_backend.rb [hunk 3]
```diff
@@ -29,7 +24,130 @@ def authentication_flow(params)
...
+        {
+          name: 'given_name',
+          value: given_name
+        },
+        {
+          name: 'family_name',
+          value: family_name
+        },
+        {
+          name: 'phone_number',
```
Comment (jakubmiarka) [inline lib/auth/authentication_backend.rb]: we haven't been using `phone_number`, any reason we're introducing it now?

File: lib/auth/authentication_backend.rb [hunk 4]
```diff
@@ -29,7 +24,130 @@ def authentication_flow(params)
...
+    client.admin_add_user_to_group(
+      user_pool_id: user_pool_id,
+      username: username,
+      group_name: group
+    )
+  rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+    raise AuthenticationBackendException.new(e.message)
+  end
+
+  def get_users(limit: 50)
```
Comment (jakubmiarka) [inline lib/auth/authentication_backend.rb]: I believe 60 is the maximum, we should use that, so we don't have to paginate for some time...
```suggestion
  def get_users(limit: 60)
```

File: lib/auth/initial_seeder.rb [hunk 5]
```diff
@@ -13,14 +17,11 @@ def initialize
...
   def gds_group_exists?
     begin
-      SelfService.service(:cognito_client).get_group(
-        group_name: TEAMS::GDS,
-        user_pool_id: Rails.configuration.cognito_user_pool_id
-      )
-    rescue Aws::CognitoIdentityProvider::Errors::ResourceNotFoundException
+      get_group(group_name: TEAMS::GDS)
+    rescue AuthenticationBackend::UserGroupNotFoundException
       Rails.logger.warn('The GDS group does not exist in Cognito!')
```
Comment (jakubmiarka) [inline lib/auth/initial_seeder.rb]: if we're abstracting it completely...
```suggestion
      Rails.logger.warn('The GDS group does not exist in the authentication backend!')
```

File: lib/auth/initial_seeder.rb [hunk 6]
```diff
@@ -36,16 +37,8 @@ def gds_group_exists?
...
-        user_pool_id: Rails.configuration.cognito_user_pool_id,
-        limit: 60
-      )
-      @gds_users = users.users.select { |user|
-        user.attributes.find { |att|
-          att.name == 'custom:roles' && att.value.include?(TEAMS::GDS)
-        }
-      }
-    rescue Aws::CognitoIdentityProvider::Errors::ServiceError => e
+      @gds_users = find_users_by_role(limit: 60, role: TEAMS::GDS)
```
Comment (jakubmiarka) [inline lib/auth/initial_seeder.rb]: if we default to 60, we don't have to pass it in here

File: lib/auth/authentication_backend.rb [hunk 7]
```diff
@@ -1,20 +1,15 @@
...
-  # Abstracted Exceptions
-  # Aws::CognitoIdentityProvider::Errors::NotAuthorizedException,
-  # Aws::CognitoIdentityProvider::Errors::UserNotFoundException,
-  # Aws::CognitoIdentityProvider::Errors::InvalidParameterException,
-  # Aws::CognitoIdentityProvider::Errors::CodeMismatchException
-  # Aws::CognitoIdentityProvider::Errors::ServiceError
-  # Aws::CognitoIdentityProvider::Errors::AliasExistsException,
-  # Aws::CognitoIdentityProvider::Errors::UsernameExistsException => e
-  # Aws::CognitoIdentityProvider::Errors::ResourceNotFoundException
   class NotAuthorizedException < StandardError; end
```
Comment (jkunle) [inline lib/auth/authentication_backend.rb]: I like how you've done this; really good stuff so far. Minor observation, it seems the convention in ruby with custom exception is to end their names in Error so maybe
 ```
class UserGroupNotFoundError < StandardError; end
 class AuthenticationBackendError < StandardError; end	 
 class AuthenticationBackendError < StandardError; end
 class UsernameExistsError < StandardError; end
 class GroupExistsError < StandardError; end
```

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -41,58 +40,31 @@ def team_valid?
...
+    add_user_to_group(username: new_user.username, group: team.name)
   end
 
   def invite_user
     begin
       team = Team.find(params[:team_id])
       invite = setup_user_in_cognito
-    rescue Aws::CognitoIdentityProvider::Errors::AliasExistsException,
-           Aws::CognitoIdentityProvider::Errors::UsernameExistsException => e
+    rescue AuthenticationBackend::UsernameExistsException => e
```
Comment (Rosa-Fox) [inline app/controllers/users_controller.rb]: How come the `AliasExistsException` has been removed?

**Pull request merged**